tokio/src/util/slab.rs

```diff
@@ -0,0 +1,745 @@
+            // It is required to avoid holding the lock when calling the
+            // provided function. The function may attempt to acquire the lock
+            // itself. If we hold the lock here while calling `f`, a deadlock
+            // situation is possible.
+            //
+            // Instead of iterating the slots directly in `page`, which would
+            // require holding the lock, the cache is updated and the slots are
+            // iterated from the cache.
+            self.cached[page_idx].refresh(&self.pages[page_idx]);
+
+            for slot_idx in 0..self.cached[page_idx].init {
+                f(self.cached[page_idx].get(slot_idx));
+            }
+        }
+    }
+
+    // Release memory back to the allocator.
+    //
+    // If pages are empty, the underlying memory is released back to the
+    // allocator.
+    pub(crate) fn compact(&mut self) {
+        // Iterate each page except the very first one. The very first page is
+        // never freed.
+        for (idx, page) in (&self.pages[1..]).iter().enumerate() {
+            let mut slots = match page.slots.try_lock() {
+                Ok(slots) => slots,
+                // If the lock cannot be acquired due to being held by another
+                // thread, don't try to compact the page.
+                _ => continue,
+            };
+
+            if slots.used > 0 || slots.slots.capacity() == 0 {
+                // The page is in use or it has not yet been allocated. Either
+                // way, there is no more work to do.
+                continue;
+            }
+
+            // Remove the slots vector from the page. This is done so that the
+            // freeing process is done outside of the lock's critical section.
+            let vec = mem::replace(&mut slots.slots, vec![]);
+            slots.head = 0;
+
+            // Drop the vec outside of the lock
+            drop(slots);
+
+            // Clear cache
+            self.cached[idx].slots = ptr::null();
+            self.cached[idx].init = 0;
+
+            drop(vec);
+        }
+    }
+}
+
+impl<T> fmt::Debug for Slab<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("Slab").finish()
+    }
+}
+
+impl<T: Entry> Allocator<T> {
+    /// Allocate a new entry and return a handle to the entry.
+    ///
+    /// Scans pages from smallest to biggest, stopping when a slot is found.
+    /// Pages are allocated if necessary.
+    ///
+    /// Returns `None` if the slab is full.
+    pub(crate) fn allocate(&self) -> Option<(Address, Ref<T>)> {
+        // Find the first available slot.
+        for page in &self.pages[..] {
+            if let Some((addr, val)) = page.allocate() {
+                return Some((addr, val));
+            }
+        }
+
+        None
+    }
+}
+
+impl<T> fmt::Debug for Allocator<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("slab::Allocator").finish()
```

Comment:
could be nice to dump some details about the state of the slab here eventually...not a big deal though.

Reply:
I improved the output.